File: inngest-spring-boot-demo/src/main/java/com/inngest/springbootdemo/UserSignupFunction.java
package com.inngest.springbootdemo;

import com.inngest.*;
import org.jetbrains.annotations.NotNull;

import java.time.Duration;
import java.util.HashMap;

public class UserSignupFunction extends InngestFunction {

    @NotNull
    @Override
    public InngestFunctionConfigBuilder config(InngestFunctionConfigBuilder builder) {
        return builder
            .id("fn-id-slug")
            .name("My Function!")
            .triggerEvent("user-signup");
    }

    @Override
    public HashMap<String, String> execute(@NotNull FunctionContext ctx, @NotNull Step step) {
        int x = 10;

        System.out.println("-> handler called " + ctx.getEvent().getName());

        int y = step.run("add-ten", () -> x + 10, Integer.class);

        Result res = step.run("cast-to-type-add-ten", () -> {
            System.out.println("-> running step 1!! " + x);
            return new Result(y + 10);
        }, Result.class);

        System.out.println("res" + res);

        step.waitForEvent("wait-for-hello", "hello", "10m", null);

        int add = step.run("add-one-hundred", () -> {
            System.out.println("-> running step 2 :) " + (res != null ? res.sum : ""));
            return (res != null ? res.sum : 0) + 100;
        }, Integer.class);

        step.sleep("wait-one-sec", Duration.ofSeconds(2));

        step.run("last-step", () -> (res != null ? res.sum : 0) * add, Integer.class);

        HashMap<String, String> data = new HashMap<String, String>() {{
            put("hello", "world");
        }};
        step.sendEvent("followup-event-id", new InngestEvent("user.signup.completed", data));

        return new HashMap<String, String>() {{
            put("message", "cool - this finished running");
        }};
    }
}

File: inngest-test-server/src/main/kotlin/com/inngest/testserver/UserSignupFunction.kt
package com.inngest.testserver

import com.inngest.*
import java.time.Duration

class ProcessUserSignup : InngestFunction() {

    override fun config(builder: InngestFunctionConfigBuilder): InngestFunctionConfigBuilder {
        return builder.id("process-user-signup")
            .triggerEvent("user-signup")
    }

    override fun execute(
        ctx: FunctionContext,
        step: Step,
    ): HashMap<String, String> {
        val x = 10

        println("-> handler called " + ctx.event.name)

        val y =
            step.run<Int>("add-ten") { ->
                x + 10
            }

        val res =
            step.run<Result>("cast-to-type-add-ten") { ->
                println("-> running step 1!! " + x)
                // throw Exception("An error!")
                Result(
                    sum = y + 10,
                )
            }
        println("res" + res)

        step.waitForEvent("wait-for-hello", "hello", "10m", "event.data.hello == async.data.hello")

        val add: Int =
            step.run<Int>("add-one-hundred") {
                println("-> running step 2 :) " + res?.sum)
                res.sum + 100
            }

        step.sleep("wait-one-sec", Duration.ofSeconds(2))

        step.run("last-step") { res.sum.times(add) ?: 0 }

        step.sendEvent("followup-event-id", InngestEvent(FOLLOW_UP_EVENT_NAME, data = hashMapOf("hello" to "world")))

        return hashMapOf("message" to "cool - this finished running")
    }
}
